File: cores/go/internal/mapper/mmc1.go
// mmc1.go - Mapper 1 (MMC1). Port of cores/c/src/mmc1.c.
package mapper

const (
	mmc1PrgBankSize = 16384
	mmc1Chr4K       = 4096
	mmc1Chr8K       = 8192
	mmc1PrgRAMSize  = 8192
	mmc1CtrlPrgMode3 uint8 = 0x0C
)

func (s *Mmc1State) prgBankCount() uint32 {
	n := s.PrgSize / mmc1PrgBankSize
	if n == 0 {
		return 1
	}
	return n
}

func (s *Mmc1State) chrBankCount() uint32 {
	n := s.ChrSize / mmc1Chr4K
	if n == 0 {
		return 1
	}
	return n
}

func (s *Mmc1State) prgMode() uint8 { return (s.Control >> 2) & 0x03 }
func (s *Mmc1State) chrMode() uint8 { return (s.Control >> 4) & 1 }

func (s *Mmc1State) prgReadBank(bank, offset uint32) uint8 {
	count := s.prgBankCount()
	bank %= count
	idx := bank*mmc1PrgBankSize + offset
	if idx >= s.PrgSize {
		return 0
	}
	return s.PrgRom[idx]
}

func (s *Mmc1State) serialWrite(addr uint16, value uint8) {
	if value&0x80 != 0 {
		s.ShiftReg = 0
		s.ShiftCount = 0
		s.Control = (s.Control & 0x13) | mmc1CtrlPrgMode3
		return
	}
	s.ShiftReg = (s.ShiftReg >> 1) | ((value & 1) << 4)
	s.ShiftCount++
	if s.ShiftCount == 5 {
		reg := uint8((addr >> 13) & 0x03)
		switch reg {
		case 0:
			s.Control = s.ShiftReg
		case 1:
			s.ChrBank0 = s.ShiftReg
		case 2:
			s.ChrBank1 = s.ShiftReg
		case 3:
			s.PrgBank = s.ShiftReg
		}
		s.ShiftReg = 0
		s.ShiftCount = 0
	}
}

func mmc1ReadPRG(m *Mapper, addr uint16) uint8 {
	s := &m.MMC1
	if addr >= 0x6000 && addr < 0x8000 {
		return s.PrgRAM[(addr-0x6000)&(mmc1PrgRAMSize-1)]
	}
	local := uint32(addr - 0x8000)
	inLow := local < mmc1PrgBankSize
	offset := local & (mmc1PrgBankSize - 1)
	switch s.prgMode() {
	case 0, 1:
		return s.prgReadBank(uint32(s.PrgBank&0x0E), local)
	case 2:
		if inLow {
			return s.prgReadBank(0, offset)
		}
		return s.prgReadBank(uint32(s.PrgBank&0x0F), offset)
	default:
		if inLow {
			return s.prgReadBank(uint32(s.PrgBank&0x0F), offset)
		}
		return s.prgReadBank(s.prgBankCount()-1, offset)
	}
}

func mmc1WritePRG(m *Mapper, addr uint16, value uint8) {
	s := &m.MMC1
	if addr >= 0x6000 && addr < 0x8000 {
		s.PrgRAM[(addr-0x6000)&(mmc1PrgRAMSize-1)] = value
		return
	}
	s.serialWrite(addr, value)
}

func mmc1ChrIndex(s *Mmc1State, addr uint16) uint32 {
	a := uint32(addr)
	count := s.chrBankCount()
	if s.chrMode() == 0 {
		bank := (uint32(s.ChrBank0 & 0x1E)) % count
		return bank*mmc1Chr4K + (a & (mmc1Chr8K - 1))
	} else if a < mmc1Chr4K {
		bank := (uint32(s.ChrBank0 & 0x1F)) % count
		return bank*mmc1Chr4K + a
	}
	bank := (uint32(s.ChrBank1 & 0x1F)) % count
	return bank*mmc1Chr4K + (a - mmc1Chr4K)
}

func mmc1ReadCHR(m *Mapper, addr uint16) uint8 {
	s := &m.MMC1
	idx := mmc1ChrIndex(s, addr)
	if idx >= s.ChrSize {
		return 0
	}
	return s.Chr[idx]
}

func mmc1WriteCHR(m *Mapper, addr uint16, value uint8) {
	s := &m.MMC1
	if !s.ChrIsRAM {
		return
	}
	idx := mmc1ChrIndex(s, addr)
	if idx < s.ChrSize {
		s.Chr[idx] = value
	}
}

func mmc1MirrorMode(m *Mapper) Mirroring {
	switch m.MMC1.Control & 0x03 {
	case 0:
		return MirrorSingleScreen0
	case 1:
		return MirrorSingleScreen1
	case 2:
		return MirrorVertical
	default:
		return MirrorHorizontal
	}
}

func Mmc1Create(m *Mapper, prg, chr []byte, mirroring Mirroring, hasBattery bool) int {
	s := &m.MMC1
	s.PrgSize = uint32(len(prg))
	if s.PrgSize > 0 {
		s.PrgRom = make([]byte, s.PrgSize)
		copy(s.PrgRom, prg)
	}
	if len(chr) == 0 {
		s.ChrSize = mmc1Chr8K
		s.Chr = make([]byte, mmc1Chr8K)
		s.ChrIsRAM = true
	} else {
		s.ChrSize = uint32(len(chr))
		s.Chr = make([]byte, s.ChrSize)
		copy(s.Chr, chr)
		s.ChrIsRAM = false
	}
	for i := range s.PrgRAM {
		s.PrgRAM[i] = 0
	}
	s.HasBattery = hasBattery
	var mirrBits uint8
	switch mirroring {
	case MirrorSingleScreen0:
		mirrBits = 0x00
	case MirrorSingleScreen1, MirrorSingleScreen2, MirrorSingleScreen3:
		mirrBits = 0x01
	case MirrorVertical:
		mirrBits = 0x02
	default:
		mirrBits = 0x03
	}
	s.Control = mmc1CtrlPrgMode3 | mirrBits
	return 0
}

func mmc1SaveState(m *Mapper, buf []byte) int {
	s := &m.MMC1
	total := mmc1StateSize(s)
	if buf == nil {
		return total
	}
	off := 0
	putU32(buf[off:], s.PrgSize); off += 4
	if s.PrgSize > 0 {
		copy(buf[off:], s.PrgRom); off += int(s.PrgSize)
	}
	putU32(buf[off:], s.ChrSize); off += 4
	buf[off] = boolByte(s.ChrIsRAM); off++
	buf[off] = boolByte(s.HasBattery); off++
	buf[off] = s.ShiftReg; off++
	buf[off] = s.ShiftCount; off++
	buf[off] = s.Control; off++
	buf[off] = s.ChrBank0; off++
	buf[off] = s.ChrBank1; off++
	buf[off] = s.PrgBank; off++
	copy(buf[off:], s.PrgRAM[:]); off += mmc1PrgRAMSize
	if s.ChrIsRAM && s.ChrSize > 0 {
		copy(buf[off:], s.Chr); off += int(s.ChrSize)
	}
	return off
}

func mmc1StateSize(s *Mmc1State) int {
	total := 4 + int(s.PrgSize) + 4 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + mmc1PrgRAMSize
	if s.ChrIsRAM {
		total += int(s.ChrSize)
	}
	return total
}

func mmc1LoadState(m *Mapper, buf []byte) bool {
	s := &m.MMC1
	if len(buf) < 4 {
		return false
	}
	off := 0
	prgSize := getU32(buf[off:]); off += 4
	if prgSize != s.PrgSize {
		return false
	}
	if s.PrgSize > 0 {
		copy(s.PrgRom, buf[off:off+int(s.PrgSize)]); off += int(s.PrgSize)
	}
	if len(buf) < off+4+8+mmc1PrgRAMSize {
		return false
	}
	chrSize := getU32(buf[off:]); off += 4
	if chrSize != s.ChrSize {
		return false
	}
	s.ChrIsRAM = buf[off] != 0; off++
	s.HasBattery = buf[off] != 0; off++
	s.ShiftReg = buf[off]; off++
	s.ShiftCount = buf[off]; off++
	s.Control = buf[off]; off++
	s.ChrBank0 = buf[off]; off++
	s.ChrBank1 = buf[off]; off++
	s.PrgBank = buf[off]; off++
	copy(s.PrgRAM[:], buf[off:off+mmc1PrgRAMSize]); off += mmc1PrgRAMSize
	if s.ChrIsRAM && s.ChrSize > 0 {
		if len(buf) < off+int(s.ChrSize) {
			return false
		}
		copy(s.Chr, buf[off:off+int(s.ChrSize)]); off += int(s.ChrSize)
	}
	return true
}

File: cores/java/src/main/java/nes/core/mappers/Mmc1.java
package nes.core.mappers;

import nes.core.Mapper;
import nes.core.Mirroring;

// Mapper 1 (MMC1). Port of cores/csharp/src/Mappers/Mmc1.cs.
public final class Mmc1 extends Mapper {
    private static final int PRG_BANK_SIZE = 16384;
    private static final int CHR_4K_SIZE = 4096;
    private static final int CHR_8K_SIZE = 8192;
    private static final int PRG_RAM_SIZE = 8192;
    private static final byte CONTROL_PRG_MODE3 = 0x0C;

    private final byte[] _prgRom;
    private final int _prgSize;
    private byte[] _chr;
    private int _chrSize;
    private boolean _chrIsRam;
    private final byte[] _prgRam = new byte[PRG_RAM_SIZE];
    private final boolean _hasBattery;
    private byte _shiftReg;
    private byte _shiftCount;
    private byte _control;
    private byte _chrBank0;
    private byte _chrBank1;
    private byte _prgBank;

    public Mmc1(byte[] prg, int prgSize, byte[] chr, int chrSize, int mirroring, boolean hasBattery) {
        mapperNum = 1;
        _prgSize = prgSize;
        _prgRom = new byte[prgSize > 0 ? prgSize : 1];
        if (prgSize > 0) System.arraycopy(prg, 0, _prgRom, 0, prgSize);
        if (chrSize == 0) {
            _chrSize = CHR_8K_SIZE;
            _chr = new byte[CHR_8K_SIZE];
            _chrIsRam = true;
        } else {
            _chrSize = chrSize;
            _chr = new byte[chrSize];
            System.arraycopy(chr, 0, _chr, 0, chrSize);
            _chrIsRam = false;
        }
        _hasBattery = hasBattery;
        _shiftReg = 0;
        _shiftCount = 0;
        byte mirrBits;
        switch (mirroring) {
            case Mirroring.SINGLE_SCREEN0: mirrBits = 0x00; break;
            case Mirroring.SINGLE_SCREEN1: mirrBits = 0x01; break;
            case Mirroring.SINGLE_SCREEN2: mirrBits = 0x01; break;
            case Mirroring.SINGLE_SCREEN3: mirrBits = 0x01; break;
            case Mirroring.VERTICAL: mirrBits = 0x02; break;
            default: mirrBits = 0x03; break;
        }
        _control = (byte) (CONTROL_PRG_MODE3 | mirrBits);
        _chrBank0 = 0;
        _chrBank1 = 0;
        _prgBank = 0;
    }

    private int prgBankCount() { return _prgSize / PRG_BANK_SIZE > 0 ? _prgSize / PRG_BANK_SIZE : 1; }
    private int chrBankCount() { return _chrSize / CHR_4K_SIZE > 0 ? _chrSize / CHR_4K_SIZE : 1; }
    private int prgMode() { return (_control >> 2) & 0x03; }
    private int chrMode() { return (_control >> 4) & 1; }

    private int prgReadBank(int bank, int offset) {
        int count = prgBankCount();
        bank %= count;
        int idx = bank * PRG_BANK_SIZE + offset;
        if (idx >= _prgSize) return 0x00;
        return _prgRom[idx] & 0xFF;
    }

    private void serialWrite(int addr, int value) {
        if ((value & 0x80) != 0) {
            _shiftReg = 0;
            _shiftCount = 0;
            _control = (byte) ((_control & 0x13) | CONTROL_PRG_MODE3);
            return;
        }
        _shiftReg = (byte) ((_shiftReg >> 1) | ((value & 1) << 4));
        _shiftCount++;
        if (_shiftCount == 5) {
            int reg = (addr >> 13) & 0x03;
            switch (reg) {
                case 0: _control = _shiftReg; break;
                case 1: _chrBank0 = _shiftReg; break;
                case 2: _chrBank1 = _shiftReg; break;
                case 3: _prgBank = _shiftReg; break;
            }
            _shiftReg = 0;
            _shiftCount = 0;
        }
    }

    @Override
    public int readPrg(int addr) {
        if (addr >= 0x6000 && addr < 0x8000)
            return _prgRam[(addr - 0x6000) & (PRG_RAM_SIZE - 1)] & 0xFF;
        int local = addr - 0x8000;
        boolean inLow = local < PRG_BANK_SIZE;
        int offset = local & (PRG_BANK_SIZE - 1);
        switch (prgMode()) {
            case 0: case 1:
                return prgReadBank(_prgBank & 0x0E, local);
            case 2:
                if (inLow) return prgReadBank(0, offset);
                return prgReadBank(_prgBank & 0x0F, offset);
            default:
                if (inLow) return prgReadBank(_prgBank & 0x0F, offset);
                return prgReadBank(prgBankCount() - 1, offset);
        }
    }

    @Override
    public void writePrg(int addr, int value) {
        if (addr >= 0x6000 && addr < 0x8000) {
            _prgRam[(addr - 0x6000) & (PRG_RAM_SIZE - 1)] = (byte) value;
            return;
        }
        serialWrite(addr, value);
    }

    private int chrIndex(int a) {
        int count = chrBankCount();
        if (chrMode() == 0) {
            int bank = (_chrBank0 & 0x1E) % count;
            return bank * CHR_4K_SIZE + (a & (CHR_8K_SIZE - 1));
        } else if (a < CHR_4K_SIZE) {
            int bank = (_chrBank0 & 0x1F) % count;
            return bank * CHR_4K_SIZE + a;
        } else {
            int bank = (_chrBank1 & 0x1F) % count;
            return bank * CHR_4K_SIZE + (a - CHR_4K_SIZE);
        }
    }

    @Override
    public int readChr(int addr) {
        int idx = chrIndex(addr);
        if (idx >= _chrSize) return 0x00;
        return _chr[idx] & 0xFF;
    }

    @Override
    public void writeChr(int addr, int value) {
        if (!_chrIsRam) return;
        int idx = chrIndex(addr);
        if (idx < _chrSize) _chr[idx] = (byte) value;
    }

    @Override
    public int mirrorMode() {
        switch (_control & 0x03) {
            case 0: return Mirroring.SINGLE_SCREEN0;
            case 1: return Mirroring.SINGLE_SCREEN1;
            case 2: return Mirroring.VERTICAL;
            default: return Mirroring.HORIZONTAL;
        }
    }

    @Override
    public boolean chrIsRam() { return _chrIsRam; }
    @Override
    public boolean hasBattery() { return _hasBattery; }

    @Override
    public int saveState(byte[] buf) {
        int total = 6 + PRG_RAM_SIZE + (_chrIsRam ? _chrSize : 0);
        if (buf == null) return total;
        int p = 0;
        buf[p++] = _shiftReg;
        buf[p++] = _shiftCount;
        buf[p++] = _control;
        buf[p++] = _chrBank0;
        buf[p++] = _chrBank1;
        buf[p++] = _prgBank;
        System.arraycopy(_prgRam, 0, buf, p, PRG_RAM_SIZE); p += PRG_RAM_SIZE;
        if (_chrIsRam && _chrSize > 0) {
            System.arraycopy(_chr, 0, buf, p, _chrSize);
            p += _chrSize;
        }
        return p;
    }

    @Override
    public boolean loadState(byte[] buf, int len) {
        int need = 6 + PRG_RAM_SIZE + (_chrIsRam ? _chrSize : 0);
        if (len < need) return false;
        int p = 0;
        _shiftReg = buf[p++];
        _shiftCount = buf[p++];
        _control = buf[p++];
        _chrBank0 = buf[p++];
        _chrBank1 = buf[p++];
        _prgBank = buf[p++];
        System.arraycopy(buf, p, _prgRam, 0, PRG_RAM_SIZE); p += PRG_RAM_SIZE;
        if (_chrIsRam && _chrSize > 0) {
            System.arraycopy(buf, p, _chr, 0, _chrSize);
            p += _chrSize;
        }
        return true;
    }
}
